rfcs/0160-portable-service-layer.md

```diff
@@ -0,0 +1,893 @@
+will automatically use a simulation strategy when a deployment scenario is
+unspecified. For example, if you only specify a `foregroundProcess` then it will
+transparently call the `daemon` tool if a process manager requires a process to
+daemonize by itself.
+
+### Process dependencies
+
+Another interesting concept is the notion process dependencies (`dependencies`
+parameter). Sometimes, a service requires the presence of another service to
+work reliably. For example, if you have a PHP application deployed to Apache
+that uses a MariaDB database for data storage, then MariaDB needs to be
+activated before Apache.
+
+The `dependencies` parameter can be used to automatically generate a
+configuration in which the order is preserved. For example, when generating
+a `systemd` configuration, this property can be translated into `Wants=` and
+`Requires=` directives. For `sysvinit` we can derived sequence numbers.
+
+In NixOS, there is no high-level concept of process dependencies. It is still
+possible to directly control the activation order by modifying systemd's `Wants=`
+and `Requires=` directives through overrides, but this is not really
+user-friendly.
+
+### Process manager-specific overrides
+
+Obviously, it is impossible to standardize all properties of all process
+managers. Sometimes, it may still be desired to specify properties that are
+unique to certain process managers.
+
+For example, `sysvinit` has the notion of run levels that are not used by other
+process managers. Specifying runlevels for `sysvinit` scripts can be done by
+defining a process-manager specific override:
+
+```nix
+{createManagedProcess, nginx, stateDir}:
+{configFile}:
+
+let
+  nginxLogDir = "${stateDir}/${instanceName}/logs";
+in
+createManagedProcess {
+  name = "nginx";
+  description = "Nginx";
+  initialize = ''
+    mkdir -p ${nginxLogDir}
+  '';
+  process = "${nginx}/bin/nginx";
+  args = [ "-c" configFile" -p" "${stateDir}/${instanceName}" ];
+
+  overrides = {
+    sysvinit = {
+      runlevels = [ 3 4 5 ];
+    };
+  };
+}
+```
+
+In the above example, the `overrides.sysvinit` property specifies
+`sysvinit`-only options. We use the override to define in which runlevels the
+script should be started.
+
+We can also use overrides for other process managers. For example, if we want
+to use the readiness-check property of systemd (which is not standardized) or
+Linux namespaces and cgroups, we can define a `systemd`-specific override.
+
+## Writing services as units of instantiation rather than singleton objects
+
+Compared to NixOS and other solutions, such as nix-darwin and system-manager,
+another feature of the Nix process management framework is that services are
+units of *instantiation* rather than *singleton* objects -- in NixOS, services
+are bound to modules. As a result, services are singleton objects that only have
+a single implementation and a single configuration.
```

Comment:
I talked to you at Nixcon about exactly this, where I even created [a gist](https://gist.github.com/infinisil/e29ee58a8184eeb7a2e139c2b2d3e4ff) for you to demonstrate how the module system can be used to support instances of services. The module system is slowly becoming the dominant way to write composable and modular Nix code, and I really think it should be used for this too. This will make integration into NixOS easier too.

Reply:
I'm going to add examples for the other two scenarios as well. Indeed, the gist that you gave me is a way to use the modular-style for multi-instance purposes.

I'd also like to explore the third option that I gave -- the hybrid approach between the functional style and the NixOS module system.

I'd like to show all three scenarios, because each of them have pros and cons. The benefit of the functional/pkgs style is that it is less memory consuming and components are configured on a need-to-know basis. The downside is obviously that you can only define one system aspect and not combine multiple aspects into one (e.g. such as adding programs to the PATH etc). Also, the NixOS module system has a "type system" which the functional style doesn't use.

The NixOS modular style is beneficial because it does fit well in the current conventions, it can be used to combine multiple system aspects into one and use of the type system to prevent certain kinds of errors. One of its major drawbacks is memory consumption/evaluation time.

The hybrid approach is something I still need to explore. Maybe it can be good a trade-off between the convenience of combining multiple aspects into one and the need-to-know properties of the functional style.

I'd like to see that we consciously pick the most suitable option out of these three.

Exploring this may take some time, maybe I'll ask you for some help @infinisil

Reply:
> I talked to you at Nixcon about exactly this [...]

... too, but I did not write a gist. My suggestion was to use `imports`.

After all, anything you can do by calling an arbitrary function, you can do by _importing into a submodule_.

Using it might look something like

```nix
# NixOS configuration.nix
{ config, lib, serviceModules, ... }:
{
  services.app-db = {
    imports = [ serviceModules.postgresql ];
    listenPort = 15432;
  };
}
```



> One of its major drawbacks is memory consumption/evaluation time.

That's mostly a scaling problem with the current (anti?)pattern of using a massive module list that's always loaded.
Using imports as above, that's solved.

A constant factor overhead does remain, but it's a very small price worth paying for a few service definitions, as opposed to, say, large package sets.

Reply:
I think I prefer something closer to @roberth's, because explicitly collecting all the instances in one place is also not terribly modular. (Feels to SQL like me, with out enough type vs table distinction.)

Reply:
> Feels to SQL like me, with out enough type vs table distinction.

Do you have a link to a page that expands on this issue? I at least am not familiar enough with SQL (or its intersection with type theory) to understand this remark.

Reply:
I do not, but perhaps i just explain it. Suppose in some language I declare a type (`struct`, whatever) called `Foo`. I can't express "give me all the `Foo`s", I have to find them myself. But in SQL I can.

Reply:
We could call this "enumerability" of instances. In the end, for the configuration to be buildable, all services have to be enumerated anyway, so I guess this is a bit about how it "feels" more than what is fundamentally possible. 
Although I think we could make this idea work by talking about the enumerability of service "subtypes" where a subtype is for example a postgresql instance (ie not talking about the other dimension, which is different supervisors).

Whether making enumeration "hard" is a good thing, I'm not sure about. For instance, it will be somewhat harder to configure a database management thing for all local databases.
Though arguably that's not a question of configuring services, but rather a constellation of services that interact. How to do this would tend to involve the specifics of the "medium" in which these services live, whether that's a NixOS configuration, or a NixOps network, or something more dynamic, etc.
Solving these as part of the (individual) service abstraction may well be a mistake. Instead, we'd have to abstract over larger systems, which seems like a different proposition altogether.

I think it'd be fine to solve reuse of fairly uninvolved service definitions before talking more about modeling larger systems of services.

Reply:
> I think it'd be fine to solve reuse of fairly uninvolved service definitions before talking more about modeling larger systems of services.

I'd say that packages with overrides in Nixpkgs are simpler to reason about than modules in NixOS, so it would be nice to have the lower implementation level to be functions, and the module-based whole function composition sit on top.

(Would also allow slightly less horribly slow to evaluate implementations of whole-system composition to share more code with NixOS)

Reply:
> Using it might look something like

I understand at a high level what you're saying but I'm not sure it's immediately clear to everyone the _details_ of what this would actually look like. I don't understand, at least.

Maybe you can lead me in the right direction. For simplicity and so that we're all clear let's pretend we're implementing this system and we happen to be implementing a `systemd` based backend - again, for simplicity and clarity using conventions we're all used to.

Back to your example. You defined an option `listenPort` in your submodule and I understand what this would look like. What I don't understand is what the implementation of this would look like - what configuration would you merge into `config` based on any `options`, like `listenPort`? In NixOS merging configuration into `config` works so well because of a number of high level abstractions defined in other modules, like `users.users`, `users.groups`, `environment.etc`, `environment.systemPackages`, and `systemd` to have a few of the most common and valuable.

In your example I don't understand how we build anything useful without those abstractions, or if we have those abstractions where they come from... unless you say each submodule `import`s everything it needs, like `systemd.nix`, `users.nix`, `environment.nix`, etc... Even if you say each submodule needs to `import` what it needs I could still use some clarification, though. Doesn't this mean a bit of redundant evaluation because multiple modules are `import`ing the same dependencies (like `systemd.nix`) over and over? If so, I guess these `import`ed files would have to be carefully designed to be used in submodules.

Any clarification you can provide would be greatly appreciated.

Reply:
@roberth maybe i'm a bit dense and this doesn't make automatic sense to me... but i was really hoping you could explain your ideas to me - sounds really interesting.

sorry for the hassle!

Reply:
> I even created [a gist](https://gist.github.com/infinisil/e29ee58a8184eeb7a2e139c2b2d3e4ff) for you to demonstrate how the module system can be used to support instances of services.

The fact that there are no examples of this technique actually being used is probably meaningful here.

@svanderburg's point remains: with Nix functions and packages you don't have to do anything special in order to "multi-instance enable" them.  It's automatic.  The gist above requires extra work in order to make a service "multi-instance aware", and comes at a cost to people who only want one instance.

We should strive for a service abstraction that makes multiple instances effortless, the way `/nix/store` makes it effortless to have multiple versions of the same package installed.

Reply:
We might end up using both something like the gist and my example because they solve different problems.

 - @infinisil's [gist](https://gist.github.com/infinisil/e29ee58a8184eeb7a2e139c2b2d3e4ff) operates at the NixOS level, allowing a "constellation" of services to be configured in one go. For example you could have a module that sets up an entire web application with local reverse proxy, database, etc, which would be counterproductive to do in a single systemd service.
 - [my example](https://github.com/NixOS/rfcs/pull/163#discussion_r1348841973) operates within a single service. This solves one of the gripes people have, that a NixOS module can affect *any* aspect of the system - no principle of least authority. In my example, the service module can only affect the system in ways that are explicitly facilitated by the system. For instance it can not add a file to `/etc` or create a user unless a proper NixOS module (ie the service abstraction layer implementation) inspects the service submodules and acts on them as it sees fit.

I imagine a crossover is also possible, where such "constellations" are defined in isolation just like individual services, but are only allotted a certain prefix of unit names for example.

Btw surely someone has come up with a better name than constellation for this.

I'll try to demonstrate the principle with a working example. (no constellations probably. that's mostly just the same pattern applied twice. interesting though. we'll see)

Reply:
> In my example, the service module can only affect the system in ways that are explicitly facilitated by the system. For instance it can not add a file to /etc or create a user unless a proper NixOS module (ie the service abstraction layer implementation) inspects the service submodules and acts on them as it sees fit.

thanks for the explanation! looking forward to seeing your example.

Reply:
Ok, did the thing. A bit rushed maybe, but a lot more complete than anything I wrote before this.

- https://github.com/NixOS/nixpkgs/pull/267111